src/main/mc_worker_pid_info.erl

```diff
@@ -117,32 +118,36 @@ install_mc_worker_info(Socket, NetModule, Database) ->
         ok
     catch
         What:Reason ->
-            NetModule:close(Socket),
             {error, {What, Reason}}
     end.
 
-detect_protocol_type(Socket, NetModule, Database) ->
-    case application:get_env(mongodb, use_legacy_protocol, auto) of
-        true -> legacy;
-        false -> op_msg; %% modern protocol based on the op_msg package
+detect_protocol_type(UseLegacyProtocol, ConnectionOpts, NetModule, Database) ->
+    case UseLegacyProtocol of
+        true ->
+            legacy;
+        false ->
+            %% modern protocol based on the op_msg package
+            op_msg;
         auto ->
-            %% Automatically detect which protocol to use. We send a
-            %% command using the old protocol. If we get back error code
-            %% 352* (UnsupportedOpQueryCommand), we are in a version that
-            %% don't support the legacy protocol so we use the op_msg based
-            %% protocol instead.
+            %% Automatically detect which protocol to use. We send a `hello' command using
+            %% the new protocol. If we have our connection closed by the server, it means
+            %% it doesn't support the new protocol.
+            %% See also:
             %% * https://github.com/mongodb/mongo/blob/5e494138af456f42381ad08748cc7fbc4ace7a60/src/mongo/base/error_codes.yml
-            Command = bson:document([{<<"notExistingCommandXyzErlang">>, <<"void">>}]),
-            Request = #'query'{
-                         collection = <<"$cmd">>,
-                         selector = Command,
-                         batchsize = -1
-                        },
-            Response = mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule),
-            case Response of
-                [{_, #reply{documents = [#{<<"code">> := 352, <<"ok">> := 0.0}|_]}}|_] ->
+            %% * https://www.mongodb.com/docs/manual/reference/command/hello/#mongodb-dbcommand-dbcmd.hello
+            %% * https://www.mongodb.com/docs/manual/reference/mongodb-wire-protocol/#std-label-wire-op-msg
+            {ok, Socket} = mc_worker_logic:connect_to_database(ConnectionOpts),
+            Command = bson:fields({hello, 1}),
+            Request = #op_msg_command{command_doc = Command, database = Database},
+            try mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule) of
+                [{_, #op_msg_response{response_doc = #{<<"ok">> := 1.0}}} | _] ->
                     op_msg;
                 _ErrorResponse ->
                     legacy
+            catch
+                _:_ ->
```

Comment:
This won’t catch the try of clause mismatch.
Change to 
```
try [{_, #op_msg_response{response_doc = #{<<"ok">> := _}}} | _] = mc_connection_man:request_raw_no_parse(Socket, Database, Request, NetModule)
```

Reply:
Sorry, which clause mismatch?  There's already a catch-all clause [here](https://github.com/emqx/mongodb-erlang/pull/39/files#diff-a1dae03dcf099259fce4c2ae4bc5c9555641c05f953a38835e35cde20e6a44a4R145).